kernel/aster-nix/src/ipc/semaphore/system_v/sem.rs

```diff
@@ -0,0 +1,307 @@
+        let status_guard = status.lock();
+        match *status_guard {
+            Status::Normal => Ok(()),
+            Status::Removed => Err(Error::new(Errno::EIDRM)),
+            Status::Pending => {
+                let mut pending_ops = if sem_op == 0 {
+                    self.pending_const.lock()
+                } else {
+                    self.pending_alters.lock()
+                };
+                pending_ops.retain(|op| op.pid != pid);
+                Err(Error::new(Errno::EAGAIN))
+            }
+        }
+    }
+
+    /// Updates pending ops after the val changed.
+    fn update_pending_ops(&self, mut val: MutexGuard<i32>) {
+        debug_assert!(*val >= 0);
+        trace!("Updating pending ops, semaphore before: {:?}", *val);
+
+        // Two steps:
+        // 1. Remove the pending_alters with `sem_op < 0` if it can.
+        // 2. If val is equal to 0, then clear pending_const
+
+        // Step one:
+        let mut pending_alters = self.pending_alters.lock();
+        pending_alters.retain_mut(|op| {
+            if *val == 0 {
+                return true;
+            }
+            // Check if the process alive.
+            if op.process.upgrade().is_none() {
+                return false;
+            }
+            debug_assert!(op.sem_buf.sem_op < 0);
+
+            if op.sem_buf.sem_op.abs() as i32 <= *val {
+                trace!(
+                    "Found removable pending op, op: {:?}, pid:{:?}",
+                    op.sem_buf.sem_op,
+                    op.pid
+                );
+
+                *val += i32::from(op.sem_buf.sem_op);
+                *self.latest_modified_pid.write() = op.pid;
+                self.update_otime();
+                *op.status.lock() = Status::Normal;
+                op.waker.wake_up();
+                false
+            } else {
+                true
+            }
+        });
+
+        // Step two:
+        if *val == 0 {
+            let mut pending_const = self.pending_const.lock();
+            pending_const.iter().for_each(|op| {
+                *op.status.lock() = Status::Normal;
+                if op.process.upgrade().is_some() {
+                    trace!("Found removable pending op, op: 0, pid:{:?}", op.pid);
+                    op.waker.wake_up();
+                }
+            });
+            pending_const.clear();
+        }
+        trace!("Updated pending ops, semaphore after: {:?}", *val);
+    }
+}
+
+pub fn sem_op(sem_id: key_t, sem_buf: SemBuf, timeout: Option<Duration>) -> Result<()> {
+    debug_assert!(sem_id > 0);
+    debug!("[semop] sembuf: {:?}", sem_buf);
+
+    let sem_sets = sem_sets();
+    let sem_set = sem_sets.get(&sem_id).ok_or(Error::new(Errno::EINVAL))?;
+    // TODO: Support permission check
+    warn!("Semaphore operation doesn't support permission check now");
+
+    let sem = sem_set
+        .get(sem_buf.sem_num as usize)
+        .ok_or(Error::new(Errno::EFBIG))?
+        .clone();
+    drop(sem_sets);
```

Comment:
We prefer using lexical lifetime so rust compiler can insert drop automatically, instead of adding `drop` manually. In your case, it can be refined as

```rust
let sem = {
    let sem_sets = sem_sets();
    let sem_set = sem_sets.get(&sem_id).ok_or(Error::new(Errno::EINVAL))?;
    // TODO: Support permission check
    warn!("Semaphore operation doesn't support permission check now");
    sem_set
        .get(sem_buf.sem_num as usize)
        .ok_or(Error::new(Errno::EFBIG))?
        .clone();
};
```